foodgroup/admin.go

```diff
@@ -0,0 +1,160 @@
+					TLVList: wire.TLVList{
+						wire.NewTLV(tag, val),
+					},
+				},
+			},
+		}
+	}
+
+	// wire.AdminTLVRegistrationStatus is used in the AIM Preferences > Privacy panel to control
+	// Allow users who know my e-mail address to find...
+	//	o Nothing about me - wire.AdminInfoRegStatusNoDisclosure
+	//	o Only that I have an account - wire.AdminInfoRegStatusLimitDisclosure
+	//	o My screen name - wire.AdminInfoRegStatusFullDisclosure
+	if _, hasRegStatus := body.TLVRestBlock.Slice(wire.AdminTLVRegistrationStatus); hasRegStatus {
+		return getAdminInfoReply(wire.AdminTLVRegistrationStatus, wire.AdminInfoRegStatusFullDisclosure), nil // todo: get from session/db
+
+	} else if _, hasEmail := body.TLVRestBlock.Slice(wire.AdminTLVEmailAddress); hasEmail {
+		return getAdminInfoReply(wire.AdminTLVEmailAddress, sess.IdentScreenName().String()+"@aol.com"), nil // todo: get from session/db
+
+	} else if _, hasNickName := body.TLVRestBlock.Slice(wire.AdminTLVScreenNameFormatted); hasNickName {
+		return getAdminInfoReply(wire.AdminTLVScreenNameFormatted, sess.DisplayScreenName().String()), nil
+
+	}
+
+	return wire.SNACMessage{
+		Frame: wire.SNACFrame{
+			FoodGroup: wire.Admin,
+			SubGroup:  wire.AdminErr,
+			RequestID: frame.RequestID,
+		},
+		Body: wire.SNACError{
+			Code: wire.ErrorCodeNotSupportedByHost,
+		},
+	}, nil
+}
+
+func (s AdminService) InfoChangeRequest(ctx context.Context, sess *state.Session, frame wire.SNACFrame, body wire.SNAC_0x07_0x04_AdminInfoChangeRequest) (wire.SNACMessage, error) {
+	var replyMessage = func(tag uint16, val any) wire.SNACMessage {
+		return wire.SNACMessage{
+			Frame: wire.SNACFrame{
+				FoodGroup: wire.Admin,
+				SubGroup:  wire.AdminInfoChangeReply,
+				RequestID: frame.RequestID,
+			},
+			Body: wire.SNAC_0x07_0x05_AdminChangeReply{
+				Permissions: wire.AdminInfoPermissionsReadWrite,
+				TLVBlock: wire.TLVBlock{
+					TLVList: wire.TLVList{
+						wire.NewTLV(tag, val),
+					},
+				},
+			},
+		}
+	}
+
+	var validateProposedName = func(name state.DisplayScreenName) (ok bool, errorCode uint16) {
+		if len(name) > 16 {
+			// proposed name is too long
+			// todo: 16 should be defined elsewhere
+			return false, wire.AdminInfoErrorInvalidNickNameLength
+		} else if name.IdentScreenName() != sess.IdentScreenName() {
+			// proposed name does not match session name (e.g. malicious client)
+			return false, wire.AdminInfoErrorInvalidNickName
+		}
+		return true, 0
+	}
+
+	if sn, hasScreenNameFormatted := body.TLVRestBlock.Slice(wire.AdminTLVScreenNameFormatted); hasScreenNameFormatted {
+		proposedName := state.DisplayScreenName(sn)
+		if ok, errorCode := validateProposedName(proposedName); !ok {
+			return wire.SNACMessage{
+				Frame: wire.SNACFrame{
+					FoodGroup: wire.Admin,
+					SubGroup:  wire.AdminErr,
+					RequestID: frame.RequestID,
+				},
+				Body: wire.SNACError{
+					Code: errorCode,
+				},
+			}, nil
+		}
+		sess.SetDisplayScreenName(proposedName)
```

Comment:
Maybe call this _after_ `UpdateDisplayScreenName()` so that the session doesn't get out of sync with the persistent store if `UpdateDisplayScreenName()` returns an error?